pkg/util/http.go

```diff
@@ -374,3 +384,46 @@ func copyValues(src url.Values) url.Values {
 func IsHTTPStatusCode(code codes.Code) bool {
 	return int(code) >= 100 && int(code) < 600
 }
+
+// RequestBuffers provides pooled request buffers.
+type RequestBuffers struct {
+	p       *sync.Pool
+	buffers []*bytes.Buffer
+	// Allows avoiding heap allocation
+	buffersBacking [10]*bytes.Buffer
+}
+
+// NewRequestBuffers returns a new RequestBuffers given a sync.Pool.
+func NewRequestBuffers(p *sync.Pool) *RequestBuffers {
+	rb := &RequestBuffers{
+		p: p,
+	}
+	rb.buffers = rb.buffersBacking[:0]
+	return rb
+}
+
+// Get obtains a buffer from the pool. It will be returned back to the pool when CleanUp is called.
+func (rb *RequestBuffers) Get(size int) *bytes.Buffer {
+	if rb == nil {
+		if size < 0 {
+			size = 0
+		}
+		return bytes.NewBuffer(make([]byte, 0, size))
+	}
+
+	b := rb.p.Get().(*bytes.Buffer)
+	b.Reset()
+	if size > 0 {
+		b.Grow(size)
+	}
+	rb.buffers = append(rb.buffers, b)
+	return b
+}
+
+// CleanUp releases buffers back to the pool.
+func (rb *RequestBuffers) CleanUp() {
+	for _, b := range rb.buffers {
+		rb.p.Put(b)
```

Comment:
@aknuds1 sorry for the late review, but we should set the `rb.buffers` values to `nil` here.

Reply:
Thanks @colega, PR coming up.

Does it have any practical effect though, since `rb` will be GC-ed after the end of the request?

Reply:
Oh, I thought it would be pooled too (if not, why we're shortening the slice below?)

Reply:
@colega I just copied it from Peter's suggestion. I figured it's for completeness.

Reply:
Hum, if we're not pooling `rb`, then we don't need to shorten the slice. If we do, then we should set the values to nil.

Reply:
It's done only to make sure that double calls to `CleanUp` don't return objects twice.

Setting it to `nil` would violate internal contract of `RequestBuffers` that `rb.buffers` uses backing array `rb.buffersBacking`. I don't think there's anything to do here.

However we could actually pool `RequestBuffers` if we wanted (do we really?).

Reply:
@colega I made a [PR](https://github.com/grafana/mimir/pull/6891) where I do as you suggest, and add tests too. PTAL.

Reply:
> It's done only to make sure that double calls to CleanUp don't return objects to the pool twice.
I see. 

> Setting rb.buffers to nil would violate internal contract of RequestBuffers that rb.buffers uses backing array rb.buffersBacking.

I don't think that should be a contract. Note that it would also be violated if we call `Get()` 11 times.

Reply:
> Setting rb.buffers to nil would violate internal contract

I interpreted @colega as meaning that items in `rb.buffers` should be nil-ed?

Reply:
> However we could actually pool RequestBuffers if we wanted (do we really?).

Pooling the `RequestBuffers` can be more efficient if we remove the individual pools: we don't need to call the synchronized sync.Pool logic for every little  buffer we need.

Reply:
>  we don't need to call the synchronized sync.Pool logic for every little buffer we need.

@colega that's a point I didn't think of.

Reply:
> I interpreted @colega as meaning that items in `rb.buffers` should be nil-ed (see PR)?

Oh, rereading Oleg's comment again, I think I misunderstood it. You're right.

Reply:
>  if we remove the individual pools

what do you mean by that?